report: adjust score's arc length accounting for rounded linecap

paulirish opened the pull request:
I noticed that with the rounded [linecap](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linecap), the arc visually extends farther than it really should.

![image](https://user-images.githubusercontent.com/39191/68079505-fc8d5f80-fda7-11e9-8a1d-d05a18f46ee3.png)

So here's the proper adjustment. After this change, the arc shortens up a little bit on both ends.

partial fix of #9803

----------------


**UPDATE.** I tweaked the math.. the arcs now go halfway between the two cases above.
![image](https://user-images.githubusercontent.com/39191/68153687-d1ffeb80-fefa-11e9-8a40-c74ff60a4ef2.png)



GIF of before and after (updated):
![Kapture 2019-11-04 at 11 59 19](https://user-images.githubusercontent.com/39191/68153568-9107d700-fefa-11e9-8723-6eda1562e120.gif)

paulirish commented on lighthouse-core/report/html/report-styles.css:
```diff
@@ -1134,7 +1133,6 @@
   text-decoration: none;
   padding: var(--score-container-padding);
 
-  --circle-border-width: 8;
```
In order to keep this `circle-stroke-width` var in CSS, we'd need to use both `requestAnimationFrame` and `getComputedStyle` in order to read it. That seemed not worth it. 

(And currently we never change this width)

patrickhulce replied:
seems like it should be relative to the viewbox specified in the svg anyhow so seems even better 👍

connorjclark commented:
wanna take care of https://github.com/GoogleChrome/lighthouse/issues/9803 too? (the part where 0 should show arc)

connorjclark commented:
the 100 is off now :(

paulirish commented:
> the 100 is off now :(

Yeah...... I don't love the 100 now.  It's gotta be a perfect ring.

patrickhulce commented:
If the end of the line weren't rounded I would agree this is the appropriate fix, but given that this change makes only the tip extend to where the line should be, it visually feels like the line does not extend far enough now because we have a grand total of like 2 pixels on the actual line.. 100 makes this obvious but a 25 too feels like it's just short of a quarter instead of being an actual quarter circle. WDYT about compensating for this fact and splitting the difference between master and this PR?

paulirish commented:
> WDYT about compensating for this fact and splitting the difference between master and this PR?

Yeah I think that's the solution. 

I think there's four total options:

* ~Trim back the arc on both ends (current impl - shown above)~
* Trim back the arc on one end and fix the 0/start alignment
* Trim back the arc on one end and fix the end alignment
* Half-trim back the arc on both ends


Also this marks the first time that trigonometry has assisted in my career. Thanks Mr Casarella!

yuinchien commented:
Love it! I think for 0, we don't need the tiny circle. 100 should be full circle yes. Looks great, thanks for the detail attention to visual!

paulirish commented:
updated.  i implemented the "Half-trim back the arc on both ends" option. 

and:

- 0 has no dot
- 1 and 2 are differentiable
- 100 is a full ring
- 99 has a satisfying amount of dimple

gif at the top is updated.

brendankenny commented:
> gif at the top is updated.

lies?

brendankenny commented on lighthouse-core/report/html/renderer/category-renderer.js:
```diff
@@ -356,6 +352,29 @@ class CategoryRenderer {
+  /**
+   * Define the score arc of the gauge
+   * Credit to xgad for the original technique: https://codepen.io/xgad/post/svg-radial-progress-meters
+   * @param {SVGCircleElement} elem
```
arcElem?

paulirish commented:
> > gif at the top is updated.
> 
> lies?

oops didn't hit submit. thx done.

brendankenny commented on lighthouse-core/report/html/renderer/category-renderer.js:
```diff
@@ -356,6 +352,29 @@ class CategoryRenderer {
+   * @param {SVGCircleElement} elem
+   * @param {number} percent
+   */
+  _setGaugeArc(elem, percent) {
```
since it's operating on the passed-in score gauge, and not a `this.gauge` or something, make this a static method?

brendankenny replied:
> since it's operating on the passed-in score gauge, and not a `this.gauge` or something, make this a static method?

OTOH, there's no other static methods in this file, and lots of them are acting more or less statically too

brendankenny commented on lighthouse-core/report/html/renderer/category-renderer.js:
```diff
@@ -356,6 +352,29 @@ class CategoryRenderer {
+   */
+  _setGaugeArc(elem, percent) {
+    const circumferencePx = 2 * Math.PI * Number(elem.getAttribute('r'));
+    // The rounded linecap of the stroke extends the arc past its start & end.
```
```suggestion
    // The rounded linecap of the stroke extends the arc past its start and end.
```
:P

brendankenny commented on lighthouse-core/report/html/renderer/category-renderer.js:
```diff
@@ -356,6 +352,29 @@ class CategoryRenderer {
+    let arcLengthPx = percent * circumferencePx - strokeWidthPx / 2;
+    // Special cases. No dot for 0, and full ring if 100
+    if (percent === 0) elem.style.opacity = '0';
+    if (percent === 1) arcLengthPx += strokeWidthPx / 2;
```
this might be clearer as 
```suggestion
    if (percent === 1) arcLengthPx = circumferencePx;
```

brendankenny commented on lighthouse-core/report/html/renderer/category-renderer.js:
```diff
@@ -356,6 +352,29 @@ class CategoryRenderer {
+    // The rounded linecap of the stroke extends the arc past its start & end.
+    // First, we tweak the -90deg rotation to adjust
+    const strokeWidthPx = Number(elem.getAttribute('stroke-width'));
+    const rotationalAdjustmentPercent = 0.25 * strokeWidthPx / circumferencePx;
```
maybe pull the `0.25` (and the `/2` below) into a constant for easy tweaking in the future?
